### PR TITLE
655 Provide grb::set( matrix, mask, value )

### LINKED ISSUE #245: Provide grb::set( matrix, value )
Same as [this variant](http://albert-jan.yzelman.net/alp/user/group__IO.html#ga1d6880056a08497b967de557b576bca5) but for Matrix.

### LINKED ISSUE #321: Provide grb::set( matrix, mask, value )

### PR DESCRIPTION
**Pull request by byjtew**
~Closes #245~
Closes #321

This MR includes the following bugfixes:
- copy-contructor of pattern matrix iterators for more than one user processes did not copy PID information correctly;
- the `fold_matrix_to_scalar` unit test relied on undefined behaviour, here fixed;
- matrix reduction into a scalar could fail if the matrix is small w.r.t. the number of user processes;
- selection operators erroneously did not appear in user documentation;
- the specification for resize and execute phases was outdated: if in execution phase it is detected that the output container has insufficient capacity, then primitives should return `ILLEGAL`;
- the BSP1D set-vector-to-value erroneously returned `FAILED` instead, herewith fixed.

### CODE REVIEW

File: include/graphblas/base/io.hpp
Comment (anyzelman) [inline include/graphblas/base/io.hpp]: This is incorrect -- if there is a matrix C with capacity zero, then running this primitive in RESIZE mode must reallocate, is it not?

File: include/graphblas/bsp1d/io.hpp
Comment (anyzelman) [inline include/graphblas/bsp1d/io.hpp:935]: The backend should define performance semantics (see other primitives for examples, should be straightforward in this case)

File: include/graphblas/base/io.hpp
Comment (anyzelman) [inline include/graphblas/base/io.hpp:1116]: Should we include a `\warning` that says on successful execution of this primitive, `C` shall be dense? (Usually, this is not what a user should want)

File: include/graphblas/bsp1d/io.hpp
Comment (anyzelman) [inline include/graphblas/bsp1d/io.hpp:951]: a static_assert for no_casting is missing

File: include/graphblas/bsp1d/io.hpp
Comment (anyzelman) [inline include/graphblas/bsp1d/io.hpp:974]: dynamic error checks seem missing: for example, a guard against overflow if m * n > 2^64-1. I think in such a case ILLEGAL should be returned, which in turn then also modifies the base specification. This may play for more primitives actually and we should open up an issue for that

File: include/graphblas/bsp1d/io.hpp
Comment (anyzelman) [inline include/graphblas/bsp1d/io.hpp:979]: Something like the following is missing:
```
const size_t full_nz = m * n; // assumes overflow is guarded against
if( phase == RESIZE ) {
    if( grb::capacity( C ) < full_nz ) {
        return grb::resize( C, full_nz );
    }
}
```

File: include/graphblas/bsp1d/io.hpp
Comment (anyzelman) [inline include/graphblas/bsp1d/io.hpp]: With the above change, this if-statement can become an assert instead (no need to consider `TRY` at the moment).

File: include/graphblas/bsp1d/io.hpp
Comment (anyzelman) [inline include/graphblas/bsp1d/io.hpp]: ALP primitives can never return `FAILED`, so this branch should never apply

File: include/graphblas/bsp1d/io.hpp
Comment (anyzelman) [inline include/graphblas/bsp1d/io.hpp:1064]: If taking the above suggestion, ret could possibly also be `ILLEGAL` to indicate overflow

File: include/graphblas/hyperdags/io.hpp
Comment (anyzelman) [inline include/graphblas/hyperdags/io.hpp]: Perf semantics are missing (just a comment that it inherits the performance semantics of the underlying backend would be sufficient)

File: include/graphblas/hyperdags/io.hpp
Comment (anyzelman) [inline include/graphblas/hyperdags/io.hpp]: I think also val should be registered as a source scalar. There should be multiple examples of that already in the hyperdags backend, for example, for folding a scalar into a vector (foldl/foldr)

File: include/graphblas/nonblocking/io.hpp
Comment (anyzelman) [inline include/graphblas/nonblocking/io.hpp]: Perf semantics should be added, otoh I'm not sure if nonblocking does that anywhere at the moment. If not, ok for me to skip, but in that case please make an issue of missing perf semantics for nonblocking)

File: include/graphblas/reference/compressed_storage.hpp
Comment (anyzelman) [inline include/graphblas/reference/compressed_storage.hpp]: This is a correct assertion, but for readability of compiler errors, the code standard is to put these static_asserts right at the beginning of each possible entry point (meaning each backend implementation of set should have a static assert like this)

File: include/graphblas/reference/io.hpp
Comment (anyzelman) [inline include/graphblas/reference/io.hpp]: Performance semantics should be given as doxygen comment here

File: include/graphblas/reference/io.hpp
Comment (anyzelman) [inline include/graphblas/reference/io.hpp]: This return-statement should be something like
```
const size_t full_nz = m * n; // and guard against overflow somewhere
if( grb::capacity( C ) < full_nz ) {
    return grb::resize( C, full_nz );
}
```

File: include/graphblas/reference/io.hpp
Comment (anyzelman) [inline include/graphblas/reference/io.hpp]: firstprivate makes unnecessary copies of const values?

File: tests/unit/setMatrixValue.cpp
Comment (anyzelman) [inline tests/unit/setMatrixValue.cpp:520]: no space after `!` (elsewhere also)

File: include/graphblas/reference/io.hpp
Comment (anyzelman) [inline include/graphblas/reference/io.hpp]: I think this sets all current values to the given one, rather than all values in the matrix. This would be the correct implementation for masked sets, in particular, for `grb::set( C, C, val )`, where the second `C` indicates the matrix itself acts as the mask for the set operation. For `grb::set( C, val )` I would expect the entire matrix `C` to have the given `val` everywhere. (Also for the CCS section ofc.)

Btw if implementing the masked variants, take care to also consider the descriptors that modify the mask behaviours, which could get tricky

**Review by anyzelman — CHANGES_REQUESTED** [review]
Also as the most major comment perhaps: should this MR not include a `grb::set( C, mask, value );`? I would expect that variant to be more useful than the fully dense case, but it may also be a confusion of what (I think) the intended semantics are and what they should be --see also the last added review comment

File: include/graphblas/reference/io.hpp
Comment (byjtew) [reply include/graphblas/reference/io.hpp]: The idea of this primitive (at leat in my head) was to assign the *existing* non-zero values to the given one. Which would be an alias for `grb::set( input/output=A, mask=A, value=x)`

**Comment by byjtew** [comment]
Goal of this MR changed form now on to a masked-set for matrix

**Comment by anyzelman** [comment]
TODOs:

- [x] Update spec
- [x] Ref impl already seems to exit
- [x] Other impl not, so need to provide
- [x] unit test needs to be extended for masked set to value

File: include/graphblas/bsp1d/io.hpp
Comment (anyzelman) [reply include/graphblas/bsp1d/io.hpp:979]: Trivial resolution: removed this entire function;)

File: include/graphblas/bsp1d/io.hpp
Comment (anyzelman) [reply include/graphblas/bsp1d/io.hpp]: see above

File: include/graphblas/bsp1d/io.hpp
Comment (anyzelman) [reply include/graphblas/bsp1d/io.hpp]: see above

File: include/graphblas/bsp1d/io.hpp
Comment (anyzelman) [reply include/graphblas/bsp1d/io.hpp:1064]: see above

File: include/graphblas/nonblocking/io.hpp
Comment (anyzelman) [reply include/graphblas/nonblocking/io.hpp]: #331 has been created

**Comment by anyzelman** [comment]
At this point, all tests OK. The remaining TODO is to check that the test for the new functionality itself, is OK;)

**Comment by anyzelman** [comment]
Collecting TODOs on this MR, suggested to handle in-order:
- [x] make the self-masked case work. The easiest is to provide a separate function that simply overwrites the values array, and leaves the other arrays alone;
- [x] make the void case work. At present, this should only entail fixing the test to handle the void case more gracefully;
- [x] the unit test should be expanded to test the structural and invert_mask descriptor;
- [x] the implementation of grb::set should call a different function for handling the structural and invert_mask cases. A separate generic function is needed since the current generic one cannot efficiently filter out such elements since it is based on raw copying of the mask's structure. This new generic function should also gracefully handle the self-masked case. Note that the spec disallows self-masking under mask inversion, so this should be possible without any intermediate buffers.

**Comment by anyzelman** [comment]
MPICH 3.3.2 detects an issue with the fold_matrix_to_scalar test, and indeed that test implementation does not conform the ALP standard. It should also be fixed-- decide later if in this issue or in a separate one.

Just now in the wee hours made the self-masked case work. This triggered some bugs here and there now also resolved.

**Comment by anyzelman** [comment]
Update: at this point, all unit tests with LPF ok except fold_matrix_to_scalar. All smoke tests with LPF OK. (MPICH 3.3.2)

**Comment by anyzelman** [comment]
> MPICH 3.3.2 detects an issue with the fold_matrix_to_scalar test, and indeed that test implementation does not conform the ALP standard. It should also be fixed-- decide later if in this issue or in a separate one.

This is confirmed fixed in this MR

**Comment by anyzelman** [comment]
All tests OK at this point (using the "difficult" MPICH version)

**Comment by anyzelman** [comment]
All tests OK at this point (using the "difficult" MPICH version)

**Comment by anyzelman** [comment]
> All tests OK at this point (using the "difficult" MPICH version)

At this point also. Now only need to increase unit test robustness and this one is done.

**Comment by anyzelman** [comment]
Running the tests again. If OK, this MR is done (the tests now generate and check for differing outputs for different descriptors)

**Comment by anyzelman** [comment]
> Running the tests again. If OK, this MR is done (the tests now generate and check for differing outputs for different descriptors)

Tests were ok but code review shows an issue that should cause the non-structural self-masked case to not work. Fixing the implementation first, and then will look at why the unit test doesn't fail because of it.

**Review by anyzelman — APPROVED** [review]
Code review OK, waiting for tests and internal CI to flag OK as well before merge.

**Comment by anyzelman** [comment]
Concept release notes:

This MR formalises the `grb::set(matrix,matrix)` and `grb::set(matrix,mask,scalar)` variants. For the former, all backends did implement it and an exhaustive unit test existed prior to this MR, but a specification of the function was missing-- this MR adds this. For the latter, all backends did implement one variant of it which was tested in the unit test for the matrix-matrix `grb::set`. This MR specifies and implements all possible variants of `grb::set(matrix,mask,scalar)`, and adds a unit test that exhaustively tests all combinations under which `grb::set(matrix,mask,scalar)` may be called.

To realise this MR, the following features were introduced:

- the ALP utilities now include a prefix-sum algorithm, both sequential and in OpenMP. The implementation of some of the new functionalities depend on this operation. Existing prefix-sum computations should rely on this new utility function; GitHub issue #332 is to follow up on that.
- ALP objects now include ALP types such as `grb::Phase`. This was required to unambiguously dispatch between `grb::set(matrix,matrix)` and `grb::set(matrix,mask,scalar)` via SFINAE.

Additionally, this MR includes the following bugfixes:

1. copy-contructor of pattern matrix iterators for more than one user processes did not copy PID information correctly;
2. the `fold_matrix_to_scalar` unit test relied on undefined behaviour, here fixed;
3. matrix reduction into a scalar could fail if the matrix is small w.r.t. the number of user processes;
4. selection operators erroneously did not appear in user documentation;
5. the specification for resize and execute phases was outdated: if in execution phase it is detected that the output container has insufficient capacity, then primitives should return `ILLEGAL`;
6. the BSP1D set-vector-to-value erroneously returned `FAILED` instead, herewith fixed.

The impact of bugs 2-4 is expected very low as it was only present in the `develop` branch and only for a limited time. Issues 1 and 5 were present longer, however, prior to crafting this MR and as far as we know, seem never to have been triggered or noticed. Issue 5 furthermore did not induce any code changes, except for the bug in issue no. 6. That bug (no. 6), however, never triggers in the currently available back-ends because vectors have fixed full capacities (the bug would only apply for final backends that do not have such a fixed full capacity).

Last, and as always, the MR minor includes code style and documentation fixes in files related and unrelated to this MR.

Many thanks to Benjamin Lozes @byjtew for drafting the initial version of the MR!

**Comment by anyzelman** [comment]
Manual unit & smoke tests with the "difficult" MPI all OK, internal CI OK, GitHub CI OK -- triggering merge=)

**Pull request merged**
